### PR TITLE
bgpv2: Introduce NoMatchingNode condition to CiliumBGPClusterConfig

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Introduce a new condition to report the error that the given node selector selects nothing. This is the first condition we introduce in the `CiliumBGPClusterConfig`. Therefore, this PR also introduces an update for the CRD and an internal infra to set conditions.

```
apiVersion: cilium.io/v2alpha1
kind: CiliumBGPClusterConfig
metadata:
  name: cilium-bgp
spec:
  bgpInstances:
  - localASN: 65000
    name: instance0
  nodeSelector:
    matchLabels:
      select: nothing
status:
  conditions:
  - lastTransitionTime: "2024-10-24T07:34:05Z"
    message: spec.nodeSelector selects nothing
    observedGeneration: 3
    reason: NoMatchingNode
    status: "True"
    type: cilium.io/NoMatchingNode
```

```release-note
bgpv2: Introduce NoMatchingNode condition to CiliumBGPClusterConfig
```

### CODE REVIEW

File: operator/pkg/bgpv2/cluster.go [hunk 1]
```diff
@@ -52,9 +59,53 @@ func (b *BGPResourceManager) reconcileBGPClusterConfig(ctx context.Context, conf
...
 }
 
+func (b *BGPResourceManager) updateConditions(config *v2alpha1.CiliumBGPClusterConfig, noMatchingNode bool) {
+	noMatchingNodeCond := meta_v1.Condition{
+		Type:               v2alpha1.BGPClusterConfigConditionNoMatchingNode,
+		Status:             meta_v1.ConditionTrue,
+		ObservedGeneration: config.Generation,
+		LastTransitionTime: meta_v1.Now(),
+		Reason:             "NoMatchingNode",
+		Message:            "spec.nodeSelector selects nothing",
```
Comment (rastislavs) [inline operator/pkg/bgpv2/cluster.go]: nit: I would maybe prefer something like `No node matches spec.nodeSelector`

File: pkg/k8s/apis/cilium.io/v2alpha1/bgp_cluster_types.go [hunk 2]
```diff
@@ -25,6 +26,11 @@ type CiliumBGPClusterConfig struct {
 
 	// Spec defines the desired cluster configuration of the BGP control plane.
 	Spec CiliumBGPClusterConfigSpec `json:"spec"`
+
+	// Status is a running status of the cluster configuration
+	//
+	// +kubebuilder:validation:Optional
+	Status CiliumBGPClusterConfigStatus `json:"status"`
```
Comment (harsimran-pabla) [inline pkg/k8s/apis/cilium.io/v2alpha1/bgp_cluster_types.go:33]: I think we need to bump CRD version as well.

**Review by harsimran-pabla — APPROVED** [review]
Thanks, great change 🚀

File: operator/pkg/bgpv2/cluster.go [hunk 3]
```diff
@@ -27,6 +29,11 @@ func (b *BGPResourceManager) reconcileBGPClusterConfigs(ctx context.Context) err
 		if rcErr != nil {
 			err = errors.Join(err, rcErr)
 		}
+
+		_, uErr := b.clientset.CiliumV2alpha1().CiliumBGPClusterConfigs().UpdateStatus(ctx, config, meta_v1.UpdateOptions{})
```
Comment (harsimran-pabla) [inline operator/pkg/bgpv2/cluster.go]: One nit: Do we want to call API server for update if nothing has changed?

Comment (harsimran-pabla) [reply operator/pkg/bgpv2/cluster.go]: If status has not mutated, imo we should not call UpdateStatus.

Comment (youngnick) [reply operator/pkg/bgpv2/cluster.go]: Yes, we should definitely check that before sending.

Comment (youngnick) [reply operator/pkg/bgpv2/cluster.go]: There's a convenience function in the `meta` library that might help: https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#SetStatusCondition

**Review by youngnick — APPROVED** [review]
Aside from checking for changes before pushing status updates, LGTM.

File: operator/pkg/bgpv2/cluster.go (hunk 3 above)
Comment (YutaroHayakawa) [reply operator/pkg/bgpv2/cluster.go]: > If status has not mutated, imo we should not call UpdateStatus.

Yeah, definitely. I'll fix that.

> There's a convenience function in the meta library that might help:

Ah, this is nice. Thanks!

**Review by rastislavs — APPROVED** [review]
Very nice, thanks!

**Comment by joestringer** [comment]
I see this was labeled as `release-note/misc`, but it's changing APIs. Should this be labeled at least as `release-note/minor`? We also may want to think about whether to increase the version (eg v1alpha1 -> v1alpha2) when making changes to the APIs. Not sure if that strictly applies when only adding new fields like status, but we have had complaints from users recently when making changes to existing APIs (even alpha ones) since users do not expect a specific version of an API to change---that's why versions exist :sweat_smile: .

**Comment by joestringer** [comment]
I think that for those users we did actually make a breaking change, that's fair. I'm not sure I understand @cilium/sig-k8s quite well enough to really determine the consequences if we're only adding new fields.

**Pull request merged**